File: LatticeKrig/src/LKDiag.f90
subroutine LKDiag(entries, nEntries, diags, nRow, nCol, matrix)
  implicit none
  integer nEntries, nRow, nCol, diags(nEntries)
  double precision entries(nEntries), matrix(nRow, nCol)
  double precision currentEntry
  integer idx, currentDiag, diagLength, diagElement

  do idx = 1, nEntries
     currentEntry = entries(idx)
     currentDiag = diags(idx)
     if (currentDiag < 0) then
       diagLength = min(nRow + currentDiag, nCol)
       do diagElement = 1,diagLength
         matrix(diagElement-currentDiag, diagElement) = currentEntry
       enddo
     else
       diagLength = min(nCol - currentDiag, nRow)
       do diagElement = 1,diagLength
         matrix(diagElement, diagElement+currentDiag) = currentEntry
       enddo
     endif
   enddo
end 

File: LatticeKrig/R/LKDiag.R
LKDiag <- function(entries, nrow, diags = NULL, ncol = nrow) {
  entries = as.double(entries)
  nEntries = as.integer(length(entries))
  nrow = as.integer(nrow)
  ncol = as.integer(ncol)
  mat <- as.double(matrix(0, nrow=nrow, ncol=ncol))
  
  if (is.null(diags)) {
    if (nEntries %% 2 == 0) {
      diags = c((-nEntries/2) : -1, 1 : (nEntries/2))
    } else {
      diags = (-(nEntries-1)/2) : ((nEntries-1)/2)
    }
  }
  
  diags = as.integer(diags)
  if(max(diags) >= ncol || min(diags) <= -nrow) {
    warning("One of the given diagonals is outside the matrix, will be ignored")
    diags = diags[(diags < ncol) && (diags > -nrow)]
  }
  if (length(entries) == 1) {
    entries = as.double(rep(entries[1], length(diags)))
    nEntries = as.integer(length(entries))
  }
  if (length(entries) != length(diags)) {
    stop("The length of entries and length of diagonals don't match")
  }
  out <- .Fortran("LKDiag", entries = entries, nEntries = nEntries, diags = diags,
                  nRow = nrow, nCol = ncol, matrix = mat, PACKAGE="LatticeKrig")
  return(matrix(out$matrix, nrow = nrow, ncol = ncol))
}
